topic-operator/src/main/java/io/strimzi/operator/topic/v2/BatchingTopicController.java

```diff
@@ -0,0 +1,1005 @@
+            hasAutoCreateTopics |= "true".equals(autoCreateTopics.value());
+        }
+        if (hasAutoCreateTopics) {
+            LOGGER.warnOp(
+                    "It is recommended that " + AUTO_CREATE_TOPICS_ENABLE + " is set to 'false' " +
+                    "to avoid races between the operator and Kafka applications auto-creating topics");
+        }
+        this.kubeClient = kubeClient;
+
+    }
+
+    /* test */ static boolean isManaged(KafkaTopic kt) {
+        return kt.getMetadata() == null
+                || kt.getMetadata().getAnnotations() == null
+                || kt.getMetadata().getAnnotations().get(MANAGED) == null
+                || !"false".equals(kt.getMetadata().getAnnotations().get(MANAGED));
+    }
+
+    private static boolean isForDeletion(KafkaTopic kt) {
+        if (kt.getMetadata().getDeletionTimestamp() != null) {
+            var deletionTimestamp = StatusUtils.isoUtcDatetime(kt.getMetadata().getDeletionTimestamp());
+            var now = Instant.now();
+            return !deletionTimestamp.isAfter(now);
+        } else {
+            return false;
+        }
+    }
+
+    static String topicName(KafkaTopic kt) {
+        String tn = null;
+        if (kt.getSpec() != null) {
+            tn = kt.getSpec().getTopicName();
+        }
+        if (tn == null) {
+            tn = kt.getMetadata().getName();
+        }
+        return tn;
+    }
+
+    static String rv(KafkaTopic kt) {
+        return kt == null || kt.getMetadata() == null ? "null" : kt.getMetadata().getResourceVersion();
+    }
+
+    private List<ReconcilableTopic> addOrRemoveFinalizer(boolean useFinalizer, List<ReconcilableTopic> reconcilableTopics) {
+        long t0 = System.nanoTime();
+        List<ReconcilableTopic> collect = reconcilableTopics.stream()
+                .map(reconcilableTopic ->
+                        new ReconcilableTopic(reconcilableTopic.reconciliation(), useFinalizer ? addFinalizer(reconcilableTopic) : removeFinalizer(reconcilableTopic), reconcilableTopic.topicName()))
+                .collect(Collectors.toList());
+        LOGGER.traceOp("{} {} topics in {}ns", useFinalizer ? "Added finalizers to" : "Removed finalizers from", reconcilableTopics.size(), System.nanoTime() - t0);
+        return collect;
+    }
+
+    private KafkaTopic addFinalizer(ReconcilableTopic reconcilableTopic) {
+        if (!reconcilableTopic.kt().getMetadata().getFinalizers().contains(FINALIZER)) {
+            LOGGER.debugCr(reconcilableTopic.reconciliation(), "Adding finalizer {}", FINALIZER);
+            long t0 = System.nanoTime();
+            KafkaTopic edit = Crds.topicOperation(kubeClient).resource(reconcilableTopic.kt()).edit(old ->
+                    new KafkaTopicBuilder(old).editOrNewMetadata().addToFinalizers(FINALIZER).endMetadata().build());
+            LOGGER.traceCr(reconcilableTopic.reconciliation(), "Added finalizer {}, took {}ns, resourceVersion now {}", FINALIZER, System.nanoTime() - t0,
```

Comment:
out of curiosity, why are you interested in how long it takes to add the finalizer?

Reply:
It's not the finalizer specifically, there is trace level logging for all communication between both Kafka and Kube. It might be better as a metric, but we don't have metrics yet (see #8650). Collecting this information _somehow_ will enable us to reason in the future about what changes might be needed to support large numbers of topics.